WordPress/Classes/Utility/Logging/WPCrashLoggingProvider.swift

```diff
@@ -23,4 +34,92 @@ class WPCrashLoggingProvider: CrashLoggingDataProvider {
 
         return TracksUser(userID: account.userID.stringValue, email: account.email, username: account.username)
     }
+
+    var loggingUploadDelegate: EventLoggingDelegate {
+        return self
+    }
+}
+
+extension WPCrashLoggingProvider: EventLoggingDelegate {
+    var shouldUploadLogFiles: Bool {
+        return
+            !ProcessInfo.processInfo.isLowPowerModeEnabled
+            && !self.userHasOptedOut
+    }
+
+    func didQueueLogForUpload(_ log: LogFile) {
+        NotificationCenter.default.post(name: WPCrashLoggingProvider.QueuedLogsDidChangeNotification, object: log)
+        DDLogDebug("📜 Added log to queue: \(log.uuid)")
+
+        if let eventLogging = CrashLogging.eventLogging {
+            DDLogDebug("📜\t There are \(eventLogging.queuedLogFiles.count) logs in the queue.")
+        }
+    }
+
+    func didStartUploadingLog(_ log: LogFile) {
+        NotificationCenter.default.post(name: WPCrashLoggingProvider.QueuedLogsDidChangeNotification, object: log)
+        DDLogDebug("📜 Started uploading encrypted log: \(log.uuid)")
+    }
+
+    func didFinishUploadingLog(_ log: LogFile) {
+        NotificationCenter.default.post(name: WPCrashLoggingProvider.QueuedLogsDidChangeNotification, object: log)
+        DDLogDebug("📜 Finished uploading encrypted log: \(log.uuid)")
+        if let eventLogging = CrashLogging.eventLogging {
+            DDLogDebug("📜\t There are \(eventLogging.queuedLogFiles.count) logs remaining in the queue.")
+        }
+    }
+
+    func uploadFailed(withError error: Error, forLog log: LogFile) {
```

Comment:
This would be a different thing, so it wouldn't change here.

The `NSError => Error` thing is less weird than it looks – `NSError` is an older generic error API that can contain a lot more information from underlying system frameworks – so casting to it lets us get additional information out of it.

It's unfortunately common and there isn't a _great_ workaround (except to use NSError everywhere, but in Swift you _usually_ just work with `Error` directly and cast into the more detailed version for special cases).

Reply:
Sounds good 👍